### PR TITLE
Runner changes - load old state, output both stream dtc & window dtc results, documentation!

### LINKED ISSUE #237: Update documentation for SparkRunner
**I certify I'm including:**

- [x] A descriptive title
- [x] A description of the problem I'm trying to solve, including *why* I think this is a problem
- [x] An overview of the suggested solution
- [x] If the feature changes current behaviour, reasons why my solution is better

### PR DESCRIPTION
**Pull request by prashastk**
#### Pull Request Checklist:
- [x] Tests for the changes have been added
- [x] Docs have been added / updated

### What kind of change does this PR introduce?
- Added support to load old state so that the same raw data need not be processed again and again.
- Output both stream dtc & window dtc results so that the old state can be persisted.
- Added documentation!

#### Link to issue
fixes #237

### CODE REVIEW

File: blurr/runner/local_runner.py [hunk 1]
```diff
@@ -29,12 +29,9 @@ def _validate_dtc_syntax(self) -> None:
 
     def _execute_for_all_identities(
             self, identity_records: Dict[str, List[Tuple[datetime, Record]]]) -> None:
```
Comment (vigbk) [inline blurr/runner/local_runner.py]: maybe we should create an alias for this List[Tuple[datetime, Record]] ?

Comment (prashastk) [reply blurr/runner/local_runner.py]: Renamed Tuple[datetime, Record] to TimeAndRecord.

File: blurr/runner/runner.py [hunk 2]
```diff
@@ -20,22 +18,27 @@
...
+    """
+    An abstract class that provides functionality to:
+    - Convert raw events in Records
+    - Process a list of Records for a user.
+
+    A class that inherits from Runner should do the following:
+    1. Call `get_per_identity_records()` using an iterator of the events available. This returns
+        a generator which creates a Tuple[Identity, TimeAndRecord]] output.
+    2. The Tuple[Identity, TimeAndRecord]] output should be grouped together by the
+         Identity to create a List of TimeAndRecord] per identity.
```
Comment (mserranom) [inline blurr/runner/runner.py]: here and in the next line, the closing square bracket of `TimeAndRecord]` is hanging.

File: blurr/runner/spark_runner.py [hunk 3]
```diff
@@ -35,44 +36,142 @@ def get_spark_session(spark_session: Optional['SparkSession']) -> 'SparkSession'
...
+        :param old_state_rdd: A previous streaming DTC state RDD as Tuple[Identity, Streaming DTC
+            State]
+        :return: RDD[Identity, Tuple[Streaming DTC State, List of Window DTC output]]
+        """
+        identity_records_with_state = identity_records
+        if old_state_rdd:
+            identity_records_with_state = identity_records.fullOuterJoin(old_state_rdd)
+        return identity_records_with_state.map(lambda x: self._execute_per_identity_records(x))
 
     def get_record_rdd_from_json_files(self,
```
[the remark this answers is not in the document]
Comment (prashastk) [reply blurr/runner/spark_runner.py]: They do have a dependency to the DTC saved as instance members. The identity and datetime is determined in the record using the stream DTC.

File: blurr/runner/runner.py (hunk 2 above)
Comment (prashastk) [reply blurr/runner/runner.py]: Fixed.

File: blurr/runner/spark_runner.py [hunk 4]
```diff
@@ -35,44 +36,142 @@ def get_spark_session(spark_session: Optional['SparkSession']) -> 'SparkSession'
...
+    def print_output(self, per_identity_data: 'RDD') -> None:
+        """
+        Basic helper function to write data to stdout. If window DTC was provided then the window
+        DTC output is written, otherwise, the streaming DTC output is written to stdout.
+
+        WARNING - For large datasets this will be extremely slow.
+
+        :param per_identity_data: Output of the `execute()` call.
+        """
+        if not self._window_dtc:
```
Comment (indradeep) [inline blurr/runner/spark_runner.py]: It's constructing documents line-by-line - why store it in memory and print - just print straight for optimal memory loading.

Comment (prashastk) [reply blurr/runner/spark_runner.py]: Discussed offline. print in lambda will be executed in the nodes which we don't want.

**Pull request merged**